### PR TITLE
Anti-features draft

### PR DESCRIPTION
**Pull request by Tagadda**
# Pitch

This is a proposal/draft to introduce anti-feature flags for apps. See https://github.com/YunoHost/issues/issues/1860 

Here, anti-feature flags work the same way categories do. Anti-features are in the `apps.json`, so we can display a warning in the catalog and display the anti-feature list before install.

From issues and the FDroid Anti-features list, here somes that can apply to a YunoHost app :

- **Tracking**: Tracks you and/or reports your activity to upstream maintainer or third parties, either without your permission or by default.
- **Non-free network services**: Promotes or depends entirely on a non-free network service.
- **Non-free addons**: Promotes other non-free applications or plugins.
- **Non-free dependencies**: Relies on software dependencies that are not free in order to run.
- **Non-free assets**: Contains and makes use of non-free assets. The most common case is apps using artwork - images, sounds, music, etc. - under a commercial license.
- **Known vulnerability**: The software presents a known security vulnerability, yet to be fixed.
- **Upstream not maintained**: This software is not maintained anymore. Expect it to break down over time, be exposed to unfixed security breaches, etc.
- **Package not maintained**: This YunoHost package is not maintained anymore.
- **Paid content**: Promotes or depends, entirely or partially, on a paid service.
- **Arbitrary limitations**: Please refer to the README.

For now, I did only a few apps but it should be a good start :

* 20euros :
    * non-free-assets : images are not bundled with the app and I couldn't find the licence for them
* alltube, invidious, youtube-dl-webui, nitter, bibliogram :
    * non-free-network
* cubiks-2048 :
    * tracking : a lots of JS from Facebook, Twitter, an image from Paypal...
* wordpress :
    * non-free-addons
    * paid-content : this one I'm not sure, I have read that it comes with Jetpack which has a paid plan)
* ztncui :
    * non-free-dependencies ?

# Todos ?

* [ ] Documentation page about anti-features
* [ ] Display anti-features in the webadmin
* [ ] Display flags on yunohost.org app list
* [ ] Anti-features in auto-generated READMEs
* [ ] Linter warning about anti-feature ? Packagers should add specifics in the READMEs ?

### CODE REVIEW

File: antifeatures.yml [hunk 1]
```diff
@@ -0,0 +1,79 @@
+- id: tracking
+  icon: user-secret
+  title:
+      en: "Tracking"
+      fr: "Pistage"
+  description:
+      en: "Track you and/or report your activity to somewhere, either without your permission or by default."
```
Comment (tituspijean) [inline antifeatures.yml:7]:
```suggestion
      en: "Tracks you and/or reports your activity to upstream maintainer or third parties, either without your permission or by default."
```

File: antifeatures.yml [hunk 2]
```diff
@@ -0,0 +1,79 @@
...
+
+- id: non-free-network
+  icon: network-wired
+  title:
+      en: "Non-free Network Services"
+      fr : "Services réseau non-libres"
+  description:
+      en: "Promote or depend entirely on a Non-Free network service."
```
Comment (tituspijean) [inline antifeatures.yml:15]:
```suggestion
      en: "Promotes or depends entirely on a non-free network service."
```

File: antifeatures.yml [hunk 3]
```diff
@@ -0,0 +1,79 @@
...
+
+- id: non-free-addons
+  icon: puzzle-piece
+  title:
+      en: "Non-free Addons"
+      fr: "Extensions non libres"
+  description:
+      en: "Promote other Non-Free applications or plugins."
```
Comment (tituspijean) [inline antifeatures.yml:23]:
```suggestion
      en: "Promotes other non-free applications or plugins."
```

File: antifeatures.yml [hunk 4]
```diff
@@ -0,0 +1,79 @@
...
+
+- id: non-free-dependencies
+  icon: book
+  title:
+      en: "Non-free dependencies"
+  description:
+      en: "Require things that are not Free Software in order to run"
```
Comment (tituspijean) [inline antifeatures.yml:30]:
```suggestion
      en: "Relies on software dependencies that are not free in order to run."
```

File: antifeatures.yml [hunk 5]
```diff
@@ -0,0 +1,79 @@
...
+      en: "Tracking"
+      fr: "Pistage"
+  description:
+      en: "Track you and/or report your activity to somewhere, either without your permission or by default."
+
+- id: non-free-network
+  icon: network-wired
+  title:
+      en: "Non-free Network Services"
+      fr : "Services réseau non-libres"
```
Comment (tituspijean) [inline antifeatures.yml:13]:
```suggestion
      en: "Promotes or relies on non-free network services."
      fr : "Promeut ou utilise des services réseau non libres."
```

File: antifeatures.yml [hunk 6]
```diff
@@ -0,0 +1,79 @@
...
+  description:
+      en: "Require things that are not Free Software in order to run"
+
+- id: non-free-assets
+  icon: file-image
+  title:
+      en: "Non-free assets"
+      fr: "Ressources non libres"
+  description:
+      en: "Contain and make use of non-ree assets. The most common case is apps using artwork - images, sounds, music, etc. - under a non-commercial license."
```
Comment (tituspijean) [inline antifeatures.yml:38]:
```suggestion
      en: "Contains and makes use of non-free assets. The most common case is apps using artwork - images, sounds, music, etc. - under a commercial license."
```

File: antifeatures.yml [hunk 7]
```diff
@@ -0,0 +1,79 @@
...
+
+- id: known-vulnerability
+  icon: face-frown
+  title:
+      en: "Known Vulnerability"
```
Comment (tituspijean) [inline antifeatures.yml:43]:
```suggestion
      en: "Known vulnerability"
```

File: antifeatures.yml [hunk 8]
```diff
@@ -0,0 +1,79 @@
...
+      fr: "Vulnérabilité connue"
+  description:
+      en: "Known security vulnerability."
```
Comment (tituspijean) [inline antifeatures.yml:46]:
```suggestion
      en: "The software presents a known security vulnerability, yet to be fixed."
```

File: antifeatures.yml [hunk 9]
```diff
@@ -0,0 +1,79 @@
...
+
+- id: deprecated-software
+  icon: face-frown
+  title:
+      en: "Upstream not maintained"
+      fr: "Application non maintenue"
+  description:
+      en: "This software is not maintained anymore."
```
Comment (tituspijean) [inline antifeatures.yml:54]:
```suggestion
      en: "This software is not maintained anymore. Expect it to break down over time, be exposed to unfixed security breaches, etc."
```

File: antifeatures.yml [hunk 10]
```diff
@@ -0,0 +1,79 @@
...
+
+- id: package-not-maintained
+  icon: face-frown
+  title:
+      en: "Package not maintained"
+      fr: "Paquet non maintenu"
+  description:
+      en: "YunoHost package is not maintained anymore."
```
Comment (tituspijean) [inline antifeatures.yml:62]:
```suggestion
      en: "This YunoHost package is not maintained anymore."
```

File: antifeatures.yml [hunk 11]
```diff
@@ -0,0 +1,79 @@
...
+
+- id: paid-content
+  icon: coins
+  title:
+      en: "Paid content"
+      fr: "Contenu payant"
+  description:
+      en: "Promote or depend entierly on a paid service."
```
Comment (tituspijean) [inline antifeatures.yml:70]:
```suggestion
      en: "Promotes or depends, entirely or partially, on a paid service."
```

File: antifeatures.yml [hunk 12]
```diff
@@ -0,0 +1,79 @@
...
+
+- id: arbitrary-limitations
+  icon: face-frown
+  title:
+      en: "Arbitrary limitations"
+      fr: "Limitations arbritraires"
```
Comment (tituspijean) [inline antifeatures.yml:76]:
```suggestion
      fr: "Limitations arbitraires"
```

**Review by tituspijean — COMMENTED** [review]
Small opinionated improvements in the descriptions. :)

File: antifeatures.yml (hunk 5 above)
Comment (Tagadda) [reply antifeatures.yml:13]: This is the `title` so it should be more concise I guess
I'll modify the description

**Comment by jedie** [comment]
Nice idea... Has somebody take a look to f-droid? 
What's about packages like minecraft server that completely non-free?!? Currently they are hidden.

**Comment by Aeris1One** [comment]
@jedie Sure thing, the list used here is based on the one, itself based on FDroid one, I posted in https://github.com/YunoHost/issues/issues/1860. This pr should only be used for discussion regarding the implementation in applist, all others discussion should happen either in the issue repo (for general discussions about the feature) or in the appropriate pr on each yunohost component (for component-specific implementation discussion).

**Comment by Tagadda** [comment]
Hey, thank you for the review!

* I'll add `replaced-by-another-app`, `alpha-software`, `not-totally-free`
* I'll replace `known_vulnerability` by `bad_security_reputation`
* I'm not sure about merging `tracking` and `non-free-assets`. I agree we need a clearer definition for `tracking`.
* Not sure about merging `not-totally-free` and `non-free-dependencies` too, i.e ztncui which is free and depends on ZeroTier which is not totally free :grin: 
* About `non-free-network` I agree we need to find an appropriate way to display that. I think anti-feature flags shouldn't "discourage too much" a user to install the app, only inform them about what the app does.
* About `paid-content` I see this more as "promotion for a paid service inside the app", and `vpnclient` does not advert any provider in particular

When I think about the way to store the tags, I think it could make more sense to have it in the app manifest, then compiled when we build the app list, and have a way to add more tags from the `apps.json` too.

**Review by yalh76 — APPROVED** [review]

**Comment by alexAubin** [comment]
Yolomerging, let's iterate later on the design once we get more experience on the battlefield :stuck_out_tongue_winking_eye:

**Pull request merged**